java/org/apache/catalina/webresources/AbstractSingleArchiveResourceSet.java

```diff
@@ -60,32 +60,42 @@ public AbstractSingleArchiveResourceSet(WebResourceRoot root, String webAppMount
         }
     }
 
+    private HashMap<String, JarEntry> loadArchiveEntries() throws IOException {
+        JarFile jarFile = null;
+        try {
+            HashMap<String,JarEntry> contents = new HashMap<>();
+            jarFile = openJarFile();
+            System.out.println("MJF ---> AbstractSingleArchiveResourceSet (" + this.getClass().getName() +
+                ") loading archive entries from jar: " + jarFile.getName());
+            Enumeration<JarEntry> entries = jarFile.entries();
+            while (entries.hasMoreElements()) {
+                JarEntry entry = entries.nextElement();
+                contents.put(entry.getName(), entry);
+            }
+            return contents;
+        } finally {
+            if (jarFile != null) {
+                closeJarFile();
+            }
+        }
+    }
 
     @Override
     protected HashMap<String,JarEntry> getArchiveEntries(boolean single) {
-        synchronized (archiveLock) {
-            if (archiveEntries == null && !single) {
-                JarFile jarFile = null;
-                archiveEntries = new HashMap<>();
-                try {
-                    jarFile = openJarFile();
-                    Enumeration<JarEntry> entries = jarFile.entries();
-                    while (entries.hasMoreElements()) {
-                        JarEntry entry = entries.nextElement();
-                        archiveEntries.put(entry.getName(), entry);
-                    }
-                } catch (IOException ioe) {
-                    // Should never happen
-                    archiveEntries = null;
-                    throw new IllegalStateException(ioe);
-                } finally {
-                    if (jarFile != null) {
-                        closeJarFile();
+        if (archiveEntries == null && !single) {
```

Comment:
```suggestion
        if (single) {
          return archiveEntries;
        }
        if (archiveEntries == null) {
```
Could we pull out the `single` check to happen upfront?  For those threads passing `single`, they will short-circuit on the upfront check.  For all stampeding threads that might obtain the lock, this simplifies the check that happens just before the lock is obtained, possibly reducing the window of contention for the lock.

Reply:
https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java

According to that Wikipedia article, we need to mark `archiveEntries` as `volatile` to make this usage correct.  Another option would be a rewrite to use a nested class, but they note there that "Performance of this implementation is not necessarily better than the `volatile` implementation."

Reply:
Hmm...this statement in that wiki bothers me: `Note the local variable "localRef", which seems unnecessary. The effect of this is that in cases where helper is already initialized (i.e., most of the time), the volatile field is only accessed once (due to "return localRef;" instead of "return helper;"), which can improve the method's overall performance by as much as 25 percent.[7]`.  Makes it sound like a volatile can be less efficient which seems counter to what we're trying to do here.  I wish there were a way I can do this without making that variable volatile...especially since it belongs to a superclass.

Reply:
https://psy-lob-saw.blogspot.com/2013/05/using-jmh-to-benchmark-multi-threaded.html says it is slower (at least it _was_ back in 2013).  But aren't we comparing this with `synchronized + file system access`?